protobuf_serialization/reader.nim

```diff
@@ -0,0 +1,295 @@
+  type T = flatType(B)
+  when E is not VarIntWrapped:
+    {.fatal: "Tried to read a VarInt without a specified encoding. This should never happen.".}
+  when sizeof(T) == 8:
+    type
+      S = int64
+      U = uint64
+  else:
+    type
+      S = int32
+      U = uint32
+
+  if key.wireType != byte(VarInt):
+    raise newException(ProtobufMessageError, "Invalid wire type for a VarInt.")
+
+  var
+    value = U(0)
+    offset = 0'i8
+    next = VAR_INT_CONTINUATION_MASK
+    preResult: T
+  while (next and VAR_INT_CONTINUATION_MASK) != 0:
+    next = stream.eofSafeRead()
+    value += (next and U(VAR_INT_VALUE_MASK)) shl offset
+    offset += 7
+
+  #Unsigned, requiring no further work.
+  when E is UIntWrapped:
+    preResult = T(value)
+  #Zig-zagged.
+  elif E is SIntWrapped:
+    preResult = T(S(value shr 1) xor -S(value and U(0b0000_0001)))
+  else:
+    #Not zig-zagged, yet negative.
+    if offset == 70:
+      #This should handle the lowest possible negative value.
+      #The cast to a signed value causes it to error/wrap to the lowest value.
+      #Said lowest value will be negative, multiplied by -1, and wrap again.
+      #This behavior requires boundChecks to be turned off in order to not raise though.
+      {.push boundChecks: off.}
+      preResult = T(-S(value + 1))
+      {.pop.}
+    #Not zig-zagged, yet positive.
+    else:
+      preResult = T(value)
+
+  #Box the result back up.
+  box(fieldVar, preResult)
+
+proc readFixed[B](
+  stream: InputStream,
+  fieldVar: var B,
+  key: byte
+) {.raises: [Defect, IOError, ProtobufEOFError, ProtobufMessageError].} =
+  type T = flatType(B)
+  when sizeof(T) == 8:
+    type U = uint64
+    if key.wireType != byte(Fixed64):
+      raise newException(ProtobufMessageError, "Invalid wire type for a Fixed64.")
+  else:
+    type U = uint32
+    if key.wireType != byte(Fixed32):
+      raise newException(ProtobufMessageError, "Invalid wire type for a Fixed32.")
+
+  var value = U(0)
+  for offset in countup(0, (sizeof(T) - 1) * 8, 8):
+    value += U(stream.eofSafeRead()) shl U(offset)
+  box(fieldVar, cast[T](value))
+
+include stdlib_readers
+
+#readValue requires readLengthDelimited function which requires readValue.
+#This would risk infinite recursion, except nested sub-buffers have a limit of 255 bytes.
+#Every sub-sub-buffer contributes to the length of the original buffer.
+proc readValueInternal[T](
+  stream: InputStream,
+  ty: typedesc[T]
+): T {.raises: [
+  Defect,
+  IOError,
+  ProtobufEOFError,
+  ProtobufMessageError
+].}
+
+proc readLengthDelimited[B](
+  stream: InputStream,
+  fieldVar: var B,
+  key: byte
+) {.raises: [
+  Defect,
+  IOError,
+  ProtobufEOFError,
+  ProtobufMessageError
+].} =
+  if key.wireType != byte(LengthDelimited):
+    raise newException(ProtobufMessageError, "Invalid wire type for a length delimited sequence/object.")
+
+  var
+    bytes = newSeq[byte](stream.eofSafeRead())
+    preResult: B
```

Comment:
After reading the length of the sequence here, you should check if the stream can buffer enough bytes with `stream.readable(n)`.

`n` should be capped to some maximum value, otherwise a remote attacker can force the code to allocate too much memory.

Once the stream has buffered enough bytes (`stream.readable(n)` returns true), you can use the `stream.withReadableRange` API to temporary restrict the readable portion of the stream before you pass it to `stdlibFromProtobuf`.

`stdlibFromProtobuf` should be actually broken into separate functions responsible for the different std lib types as indicated in another comment. Each such function should take an `InputStream` instead of `seq[byte]` and it will consume the data naturally with all the other existing `readValue` functions that work with streams already. No need for byte sequence allocations and `unsafeMemoryInput` as done below.